Source/MockCapture.cs

```diff
@@ -0,0 +1,139 @@
+//OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+//SUCH DAMAGE.
+
+//[This is the BSD license, see
+// http://www.opensource.org/licenses/bsd-license.php]
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace Moq
+{
+    /// <summary>
+    /// Captures mock parameters.
+    /// </summary>
+    /// <example>
+    /// Arrange code:
+    /// <code>
+    /// var capture = new MockCapture{string}();
+    /// mock.Setup(x => x.DoSomething(capture.CaptureAny()));
+    /// </code>
+    /// Assert code:
+    /// <code>
+    /// Assert.Equal("Hello!", capture.Last);
+    /// </code>
+    /// </example>
+    public class MockCapture<T>
+    {
+        private readonly List<T> capturedValues = new List<T>();
+        private Action<T> callback;
+        private Func<T, T> copier;
+
+        /// <summary>
+        /// Initializes an instance of the capture.
+        /// </summary>
+        public MockCapture()
+        {
+            this.All = this.capturedValues.AsReadOnly();
+        }
+
+        /// <summary>
+        /// Gets all the captured parameter values.
+        /// </summary>
+        public IReadOnlyList<T> All { get; private set; }
+
+        /// <summary>
+        /// Gets the last captured parameter value.
+        /// </summary>
+        public T Last { get { return this.All.Last(); } }
+        
+        /// <summary>
+        /// Gets the single captured parameter value.
+        /// </summary>
+        public T Single { get { return this.All.Single(); } }
+
+        /// <summary>
+        /// Gets the parameter to use in the setup expression.
+        /// </summary>
+        /// <returns>The setup expression parameter</returns>
+        public T CaptureAny()
+        {
+            return Match.Create(value => this.CaptureParameter(value), () => It.IsAny<T>());
+        }
+
+        /// <summary>
+        /// Gets the parameter to use in the setup expression.
+        /// </summary>
+        /// <returns>The setup expression parameter</returns>
+        public T Capture(Expression<Func<T, bool>> match)
+        {
+            var matchDelegate = match.Compile();
+            return Match.Create(value => matchDelegate.Invoke(value) && this.CaptureParameter(value), () => It.Is<T>(match));
+        }
+
+        /// <summary>
+        /// Copy captured parameters with the specified delegate.
+        /// </summary>
+        /// <param name="copier">The delegate used to clone parameters</param>
+        public void CopyOnCapture(Func<T, T> copier)
+        {
+            this.copier = copier;
+        }
+
+        /// <summary>
+        /// Provide an action to run on captured parameters.
+        /// </summary>
+        /// <paramref name="callback">an action to run on the captured parameters</paramref>
+        public void RunOnCapture(Action<T> callback)
+        {
+            this.callback = callback;
+        }
+
+        private bool CaptureParameter(T item)
```

Comment:
I like your idea to use a constructor parameter to provide the copy or capture callback.

However in my experience 99% of the time the developer will need to access the list of the captured objects or the last captured object. And because this is the main use case, it should be very straightforward. If the captured object list is removed from the MockCapture class, we must provide an easy way to create a capture from a list.

What I can think up at the moment:
- Add an extension method for `ICollection<T>`
```cs
mock.Setup(x => x.DoSomething(items.AsCapture());
```
- Add an other class, for example `MockCaptureList<T>` that exposes the captured items

But in my opinion it is simpler to expose the captured items on the `MockCapture` class.